parlai/core/metrics.py

```diff
@@ -307,39 +335,35 @@ def update(self, observation, labels):
                 self.metrics['correct_cnt'] += 1
 
             # F1 and BLEU metrics.
-            f1 = _f1_score(prediction, labels)
-            bleu = _bleu(prediction, labels)
-            rouge1, rouge2, rougel = _rouge(prediction, labels)
+            if 'f1' in self.metrics_list:
+                f1 = _f1_score(prediction, labels)
+            if 'bleu' in self.metrics_list:
+                bleu = _bleu(prediction, labels)
+            if self.compute_rouge_l or self.compute_rouge_n > 0:
+                rougen, rougel = _rouge(prediction, labels, self.compute_rouge_n, self.compute_rouge_l)
 
             with self._lock():
-                self.metrics['f1'] += f1
-                self.metrics['f1_cnt'] += 1
-                if bleu is not None:
+                if 'f1' in self.metrics:
+                    self.metrics['f1'] += f1
+                    self.metrics['f1_cnt'] += 1
+                if 'bleu' in self.metrics:
                     self.metrics['bleu'] += bleu
                     self.metrics['bleu_cnt'] += 1
-                if rouge1 is not None:
-                    self.metrics['rouge-1'] += rouge1
-                    self.metrics['rouge-2'] += rouge2
+                if self.compute_rouge_l:
                     self.metrics['rouge-L'] += rougel
-                    self.metrics['rouge-1_cnt'] += 1
-                    self.metrics['rouge-2_cnt'] += 1
                     self.metrics['rouge-L_cnt'] += 1
+                if self.compute_rouge_n:
+                    for idx in range(1, self.compute_rouge_n):
+                        self.metrics['rouge-{}'.format(idx)] += rougen[idx - 1]
+                        self.metrics['rouge-{}_cnt'.format(idx)] += 1
 
         # Ranking metrics.
         self._update_ranking_metrics(observation, labels)
 
         # User-reported metrics
         if 'metrics' in observation:
             for k, v in observation['metrics'].items():
-                if k not in [
-                    'correct',
-                    'f1',
-                    'hits@k',
-                    'bleu',
-                    'rouge-1',
-                    'rouge-2',
-                    'rouge-L',
-                ]:
+                if k not in __ALL__METRICS__:
```

Comment:
if we keep this condition we might need to make all metrics a set instead I think.